### PR TITLE
Added multi-threaded downloads/uploads in s3

### PR DESCRIPTION
**Pull request by agrawalsourav98**
Added multi threaded downloads and uploads for s3. Improves download/upload speed considerably as a single file wouldn't be able to maximize bandwidth usage.

Also, added recursive behavior in download/upload.

I have been using multi-threading in the solar ortho detection package and as I am moving things to production and planning the use this repo for most of the interactions with s3, I thought I would add this here.

### CODE REVIEW

File: ml_dronebase_data_utils/s3/s3.py [hunk 1]
```diff
@@ -134,21 +137,35 @@ def upload_dir(local_path: str, s3_url: str, exist_ok: bool = True):
...
-        if exist_ok:
-            try:
-                client.head_object(Bucket=bucket_name, Key=s3_path)
-            except ClientError:
+
+    lock = threading.Lock()
+
+    with tqdm(total=num_files) as pbar:
+
+        def upload_file_local(filename):
```
Comment (agrawalsourav98) [inline ml_dronebase_data_utils/s3/s3.py:148]: Moved code into a local scoped function and use of separate sessions for thread-safe operations as advised by boto3.

File: ml_dronebase_data_utils/s3/s3.py [hunk 2]
```diff
@@ -189,20 +206,37 @@ def download_dir(
...
-        if size_limit is not None:
-            if obj.size > size_limit:
-                continue
-        bucket.download_file(obj.key, target)
+
+    lock = threading.Lock()
+
+    with tqdm(total=num_objects) as pbar:
+
+        def download_file_local(obj):
```
Comment (agrawalsourav98) [inline ml_dronebase_data_utils/s3/s3.py:214]: Moved code into a local function

File: ml_dronebase_data_utils/s3/s3.py [hunk 3]
```diff
@@ -242,6 +276,22 @@ def split_dataset(
         _split_unlabeled_dataset(data_url, train_split, val_split)
 
 
+def move_file(bucket: str, prefix: str, file: str):
```
Comment (agrawalsourav98) [inline ml_dronebase_data_utils/s3/s3.py:279]: New move file function

File: ml_dronebase_data_utils/s3/s3.py [hunk 4]
```diff
@@ -250,11 +300,9 @@ def move_files(bucket: str, prefix: str, files: List[str]):
         prefix (str): prefix to move the files to.
         files (List[str]): list of files being moved.
     """
-    s3 = boto3.resource("s3")
-    for file in files:
-        copy_source = {"Bucket": bucket, "Key": file}
-        new_prefix = os.path.join(prefix, os.path.basename(file))
-        s3.meta.client.copy(copy_source, bucket, new_prefix)
+    # Multi-threaded move for faster operations
+    with ThreadPool(processes=64) as pool:
```
Comment (agrawalsourav98) [inline ml_dronebase_data_utils/s3/s3.py:304]: Multi-threaded move operation

File: ml_dronebase_data_utils/s3/s3.py [hunk 5]
```diff
@@ -134,21 +140,44 @@ def upload_dir(local_path: str, s3_url: str, exist_ok: bool = True):
         exist_ok (bool): Decides whether or not to ignore existing files. Default True.
     """
     bucket_name, prefix = _parse_url(s3_url)
-    client = boto3.client("s3")
 
-    files = os.listdir(local_path)
+    # Allow recursive uploads and handle IsDirectoryError
+    files = glob(f"{local_path}/**", recursive=recursive)
```
Comment (agrawalsourav98) [inline ml_dronebase_data_utils/s3/s3.py:145]: Added recursive behaviour

**Comment by isaaccorley** [comment]
@agrawalsourav98 Have you tried using the sync_dir function? This just uploads/downloads a directory to s3 using aws s3 sync which parallelizes uploads already. This seems to be what you want already.

**Comment by agrawalsourav98** [comment]
> @agrawalsourav98 Have you tried using the sync_dir function? This just uploads/downloads a directory to s3 using aws s3 sync which parallelizes uploads already. This seems to be what you want already.

My main goal was to add it to the `move_files` function but I saw the other functions so just updated them.

**Comment by isaaccorley** [comment]
@agrawalsourav98 It seems you merged this but nobody approved it? Also you are hardcoding the number of threads used. It would be better to have this as an input argument instead.

**Comment by agrawalsourav98** [comment]
> @agrawalsourav98 It seems you merged this but nobody approved it? Also you are hardcoding the number of threads used. It would be better to have this as an input argument instead.

Sorry for merging without approval but I wanted to use this and it didn't get approved yesterday.

I have experimented with different values and found that 64 is a good number for maximizing bandwidth without overloading things. Do you foresee any use cases for user defined number of processes?

**Comment by isaaccorley** [comment]
Assuming every user has the same hardware as you is not a reason to hardcode values. It would be better to use 64 as a default value to an input argument.

**Comment by agrawalsourav98** [comment]
Noted. I would make it as an adjustable parameter.

**Pull request merged**